### ISSUE TITLE
关于测试速度

### ISSUE THREAD

**Issue by ylc2580** [cfzd/Ultra-Fast-Lane-Detection#20]
你好，咱们的论文中说在1080ti上测试Res34的时间为5.9ms，Res18的时间为3.2ms，但我这边用咱们默认的demo测试（测试模块见下面）的话，时间接近100ms，是咱们的测试代码需要修改还是？
start = time.clock()
with torch.no_grad():
        out = net(imgs)
end__ = time.clock()
print("time spend = ",end__ - start)

**Comment by cfzd** [comment]
@ylc2580 
Hi, I tested a similar code as yours.
```
imgs, names = data
imgs = imgs.cuda()
t1 = time.clock()
with torch.no_grad():
    out = net(imgs)
t2 = time.clock()
print('Elapsed time: ', t2-t1)
```
Here is the output I got:
```
python demo.py ./configs/culane.py --test_model ./culane_18.pth 
merge  test_model  config
start testing...
test0_normal.avi
demo.py:58: DeprecationWarning: time.clock has been deprecated in Python 3.3 and will be removed from Python 3.8: use time.perf_counter or time.process_time instead
  t1 = time.clock()
demo.py:61: DeprecationWarning: time.clock has been deprecated in Python 3.3 and will be removed from Python 3.8: use time.perf_counter or time.process_time instead
  t2 = time.clock()
Elapsed time:  0.36584599999999945
Elapsed time:  0.003686000000000078
Elapsed time:  0.003525999999999918
Elapsed time:  0.0034369999999999123
Elapsed time:  0.004587000000000785
Elapsed time:  0.003294999999999604
Elapsed time:  0.004569000000000045
Elapsed time:  0.0033309999999993067
Elapsed time:  0.003303000000000722
Elapsed time:  0.003294000000000352
Elapsed time:  0.003281999999999563
Elapsed time:  0.0032639999999997116
Elapsed time:  0.003270999999999802
```


Besides the above `demo.py` code, you can directly run `speed.py` to test the speed.


If you get 100ms inference speed, I think it might be these problems:
- the program is not running in a GPU
- the backbone is not Res18 or Res34
- the batch size is not 1
- `torch.backends.cudnn.benchmark = True` is not enabled
- GPU is occupied by other programs

**Comment by ylc2580** [comment]
> @ylc2580
> Hi, I tested a similar code as yours.
> 
> ```
> imgs, names = data
> imgs = imgs.cuda()
> t1 = time.clock()
> with torch.no_grad():
>     out = net(imgs)
> t2 = time.clock()
> print('Elapsed time: ', t2-t1)
> ```
> 
> Here is the output I got:
> 
> ```
> python demo.py ./configs/culane.py --test_model ./culane_18.pth 
> merge  test_model  config
> start testing...
> test0_normal.avi
> demo.py:58: DeprecationWarning: time.clock has been deprecated in Python 3.3 and will be removed from Python 3.8: use time.perf_counter or time.process_time instead
>   t1 = time.clock()
> demo.py:61: DeprecationWarning: time.clock has been deprecated in Python 3.3 and will be removed from Python 3.8: use time.perf_counter or time.process_time instead
>   t2 = time.clock()
> Elapsed time:  0.36584599999999945
> Elapsed time:  0.003686000000000078
> Elapsed time:  0.003525999999999918
> Elapsed time:  0.0034369999999999123
> Elapsed time:  0.004587000000000785
> Elapsed time:  0.003294999999999604
> Elapsed time:  0.004569000000000045
> Elapsed time:  0.0033309999999993067
> Elapsed time:  0.003303000000000722
> Elapsed time:  0.003294000000000352
> Elapsed time:  0.003281999999999563
> Elapsed time:  0.0032639999999997116
> Elapsed time:  0.003270999999999802
> ```
> 
> Besides the above `demo.py` code, you can directly run `speed.py` to test the speed.
> 
> If you get 100ms inference speed, I think it might be these problems:
> 
> * the program is not running in a GPU
> * the backbone is not Res18 or Res34
> * the batch size is not 1
> * `torch.backends.cudnn.benchmark = True` is not enabled
> * GPU is occupied by other programs
> * the result of 100ms is obtained when the program is not stable (the first inference is always slow)

你好，我的测试命令为：python3.7 demo_ylc.py configs/culane.py --test_model culane_18.pth 
测试代码为：
import torch, os, cv2
from model.model import parsingNet
from utils.common import merge_config
from utils.dist_utils import dist_print
import torch
import scipy.special, tqdm
import numpy as np
import torchvision.transforms as transforms
from data.dataset import LaneTestDataset
from PIL import Image




if __name__ == "__main__":
    torch.backends.cudnn.benchmark = True
    args, cfg = merge_config()
    dist_print('start testing...')
    assert cfg.backbone in ['18','34','50','101','152','50next','101next','50wide','101wide']

    if cfg.dataset == 'CULane':
        cls_num_per_lane = 18
    elif cfg.dataset == 'Tusimple':
        cls_num_per_lane = 56
    else:
        raise NotImplementedError

    net = parsingNet(pretrained = False, backbone=cfg.backbone,cls_dim = (cfg.griding_num+1,cls_num_per_lane,4),
                    use_aux=False).cuda() # we dont need auxiliary segmentation in testing

    state_dict = torch.load(cfg.test_model, map_location='cpu')['model']
    compatible_state_dict = {}
    for k, v in state_dict.items():
        if 'module.' in k:
            compatible_state_dict[k[7:]] = v
        else:
            compatible_state_dict[k] = v

    net.load_state_dict(compatible_state_dict, strict=False)
    net.eval()

    img_transforms = transforms.Compose([
        transforms.Resize((288, 800)),
        transforms.ToTensor(),
        transforms.Normalize((0.485, 0.456, 0.406), (0.229, 0.224, 0.225)),
    ])

    
    cap = cv2.VideoCapture("/media/CarDataSet/p9000/video_all/0134.mp4")
    fourcc = cv2.VideoWriter_fourcc(*'MJPG')
    rval,frame = cap.read()
    frame = frame[490:1080,0:1640,:]
    vout = cv2.VideoWriter(str(111)+'.avi', fourcc , 30.0, (frame.shape[1],frame.shape[0]))

    print("w = {},h = {}".format(cap.get(3),cap.get(4)))
    import time
    

    print('加载cuda是否成功：',torch.cuda.is_available())
    framenmber = 0
    while 1:
        rval,frame = cap.read()
        # print("framenumber = ",framenmber," frame shape = ",frame.shape)
        

        if rval == False:
            break
        frame = frame[490:1080,0:1640,:]
        # cv2.imshow(" ",frame)
        # cv2.waitKey(5)
        img  = cv2.cvtColor(frame,cv2.COLOR_BGR2RGB)
        img_ = Image.fromarray(img)
        imgs = img_transforms(img_)
        imgs = imgs.unsqueeze(0)
        imgs = imgs.cuda()

        start = time.clock()
        with torch.no_grad():
            out = net(imgs)
        end__ = time.clock()
        print("time spend = ",end__ - start)

        col_sample = np.linspace(0, 800 - 1, cfg.griding_num)
        col_sample_w = col_sample[1] - col_sample[0]

        out_j = out[0].data.cpu().numpy()
        out_j = out_j[:, ::-1, :]
        prob = scipy.special.softmax(out_j[:-1, :, :], axis=0)
        idx = np.arange(cfg.griding_num) + 1
        idx = idx.reshape(-1, 1, 1)
        loc = np.sum(prob * idx, axis=0)
        out_j = np.argmax(out_j, axis=0)
        loc[out_j == cfg.griding_num] = 0
        out_j = loc

        for i in range(out_j.shape[1]):
            if np.sum(out_j[:, i] != 0) > 2:
                for k in range(out_j.shape[0]):
                    if out_j[k, i] > 0:
                        ppp = (int(out_j[k, i] * col_sample_w * frame.shape[1] / 800) - 1, int(frame.shape[0] - k * 20) - 1)
                        cv2.circle(frame,ppp,5,(0,255,0),-1)
        vout.write(frame)
        framenmber+=1
    vout.release()

打印结果为：
加载cuda是否成功： True
time spend =  0.0027869999999836637
time spend =  0.07567800000001057
time spend =  0.09471199999998703
time spend =  0.047696999999999434
time spend =  0.09882100000001515
time spend =  0.04665499999998701
time spend =  0.08185199999999782
time spend =  0.002789000000007036
time spend =  0.08306300000000988
time spend =  0.002793999999994412
time spend =  0.046712999999982685
time spend =  0.0741480000000081
time spend =  0.003150000000005093
time spend =  0.09956399999998666
time spend =  0.051478999999972075
time spend =  0.09497700000000009
time spend =  0.003012000000012449
time spend =  0.0029989999999884276
time spend =  0.005378000000007432
time spend =  0.09461699999999951
time spend =  0.08248599999998873
time spend =  0.09472600000000853
time spend =  0.06109099999997625
time spend =  0.002759000000025935
time spend =  0.002806999999990012
time spend =  0.0026730000000156906
time spend =  0.057103000000012116
time spend =  0.09465000000000146
time spend =  0.05491399999999658
time spend =  0.09452399999997851
time spend =  0.063508000000013
time spend =  0.09651300000001584
time spend =  0.002640999999982796
time spend =  0.09144200000000069
time spend =  0.004300000000000637
time spend =  0.06152900000000727
time spend =  0.0860230000000115
time spend =  0.09397199999997952
time spend =  0.002645999999998594
time spend =  0.08176199999999767
time spend =  0.005199000000004617
time spend =  0.0614309999999989
time spend =  0.005011999999993577
time spend =  0.0026340000000004693
time spend =  0.07338499999997339
time spend =  0.0026569999999992433
time spend =  0.06342600000002108
time spend =  0.09458800000001588
time spend =  0.0530289999999809
time spend =  0.07507499999999823
time spend =  0.0913630000000012
time spend =  0.0027680000000032123
time spend =  0.09511999999998011
time spend =  0.08909599999998363
time spend =  0.06627000000000294
time spend =  0.0028160000000241325
time spend =  0.04800900000000752
time spend =  0.0656409999999994
time spend =  0.08650199999999586
time spend =  0.09449499999999489
time spend =  0.07666100000000142
time spend =  0.0807709999999986
time spend =  0.09570300000001453
time spend =  0.003033999999985326
time spend =  0.0812750000000051
time spend =  0.0030220000000156233
time spend =  0.06073200000000156
time spend =  0.09497999999999251
time spend =  0.0030829999999753
time spend =  0.09640099999998597
time spend =  0.061208999999990965
time spend =  0.07181200000002264
time spend =  0.0684840000000122
time spend =  0.058485999999987826
time spend =  0.0861099999999908
time spend =  0.05759800000001292
time spend =  0.09539900000001467
time spend =  0.06328600000000506
time spend =  0.09527199999999425
time spend =  0.06137999999998556
time spend =  0.09482900000000427
time spend =  0.09486499999999864
time spend =  0.07418800000002079
time spend =  0.08596499999998741
time spend =  0.09492000000000189
time spend =  0.09478899999999157
time spend =  0.0030050000000017008
time spend =  0.003012000000012449
time spend =  0.0030859999999961474
time spend =  0.09467700000001855
time spend =  0.0034050000000149794
time spend =  0.07837500000002251
time spend =  0.07457400000001257
有时候可以达到3ms，但大多数是超过3的，
还有就是想问一下，咱们的数据集是怎么标注的呢？用的啥软件呢？我看下载的数据集是图片和txt的，不是很理解怎么操作的，谢谢啦

**Comment by cfzd** [comment]
@ylc2580 
If you get one image running at 3ms, you should also get other images running at 3ms. There is no difference in these images when running the testing.

So I think it might be your hardware problem. You can check the device using `nvidia-smi` when running. 

As for datasets, CULane and Tusimple are public datasets released by other researchers or institutes. We did not annotate these datasets. So I don't know how the datasets are annotated. More information can be found in [here](https://xingangpan.github.io/projects/CULane.html) and [here](https://github.com/TuSimple/tusimple-benchmark).

**Comment by ylc2580** [comment]
大佬你好，
（1）我用咱们里面的speed.py测试，速度基本恒定在2.7ms左右（和你上面的3.5ms左右的差距应该是在显卡上），很稳定，显存占用1100左右，另外，为了排除生成的全0图像和实质图像的区别，全1图像和随机生成的图像也试过了。都很稳定。约2.7ms。
（2）测试视频过程中，如上面所述，我的torch.cuda.is_available()返回值为true，而且watch -n 1 nvidia-smi中代码显存占用率约为1100左右（和（1）基本一致），但速度波动很大，3ms-90ms，这两个的对比应该排除硬件问题，请问咱们有无跑过视频或者或者其他数据呢？

**Comment by cfzd** [comment]
@ylc2580 
As you may see this log
> Elapsed time:  0.36584599999999945
Elapsed time:  0.003686000000000078
Elapsed time:  0.003525999999999918
Elapsed time:  0.0034369999999999123
Elapsed time:  0.004587000000000785
Elapsed time:  0.003294999999999604
Elapsed time:  0.004569000000000045
Elapsed time:  0.0033309999999993067
Elapsed time:  0.003303000000000722
Elapsed time:  0.003294000000000352
Elapsed time:  0.003281999999999563
Elapsed time:  0.0032639999999997116
Elapsed time:  0.003270999999999802

This is the log I got when generating the video. 

I think the problem might be the `time.clock` function, which ***only*** counts ***CPU clocks***. But the code is running at ***GPU***. Theoretically, the elapsed time of `time.clock` should less than the one of `time.time`, but it is not true in your case. I don't know the reason for your problem, but it should be the problem of `time.clock`.
